File: forms/otolaryngology-assessment/front-end-form-with-html/js/snot22-grader.js
// SNOT-22 grader. Pure functions: take an `AssessmentData`, return total
// (0-110), severity level, answered count, and per-question audit trail.
//
// Severity bands:
//   0  - 7  -> Mild     (minimal impact on daily life)
//   8  - 19 -> Moderate (moderate impact; targeted management advised)
//   >= 20  -> Severe   (significant impact; consider specialist referral)

/**
 * @typedef {import('./types.js').AssessmentData} AssessmentData
 * @typedef {import('./types.js').FiredRule} FiredRule
 * @typedef {import('./types.js').SeverityLevel} SeverityLevel
 */

(function () {
'use strict';
window.OtolaryngologyAssessment = window.OtolaryngologyAssessment || {};
const { snot22Rules } = window.OtolaryngologyAssessment;

/** @returns {SeverityLevel} */
function classifySnot22Score(score) {
  if (score >= 20) return 'severe';
  if (score >= 8) return 'moderate';
  return 'mild';
}

function severityLabel(level) {
  switch (level) {
    case 'severe': return 'Severe — significant impact';
    case 'moderate': return 'Moderate — targeted management advised';
    case 'mild': return 'Mild — minimal impact';
    default: return '';
  }
}

function severityClass(level) {
  switch (level) {
    case 'severe': return 'severity-severe';
    case 'moderate': return 'severity-moderate';
    case 'mild': return 'severity-mild';
    default: return '';
  }
}

/**
 * Sum all 22 SNOT items the patient answered, classify the result.
 * @param {AssessmentData} data
 */
function calculateSnot22(data) {
  /** @type {FiredRule[]} */
  const firedRules = [];
  let totalScore = 0;
  let answeredCount = 0;

  for (const rule of snot22Rules) {
    try {
      const { score, answered } = rule.evaluate(data);
      if (answered) {
        answeredCount++;
        totalScore += score;
        firedRules.push({
          id: rule.id,
          category: rule.category,
          description: rule.description,
          score
        });
      }
    } catch (e) {
      console.warn(`SNOT-22 rule ${rule.id} evaluation failed:`, e);
    }
  }

  const severityLevel = classifySnot22Score(totalScore);
  return { totalScore, severityLevel, answeredCount, firedRules };
}

Object.assign(window.OtolaryngologyAssessment, {
  classifySnot22Score,
  severityLabel,
  severityClass,
  calculateSnot22
});
})();

File: forms/otolaryngology-assessment/full-stack-with-loco-tera-htmx-alpine/src/engine/snot22_grader.rs
use super::flagged_issues::detect_additional_flags;
use super::snot22_rules::all_rules;
use super::types::{AssessmentData, FiredRule, GradingResult, SeverityLevel};

/// Classify a SNOT-22 total into mild / moderate / severe.
///
/// - 0 - 7  -> `mild`
/// - 8 - 19 -> `moderate`
/// - >= 20  -> `severe`
pub fn classify_snot22_score(score: i32) -> SeverityLevel {
    if score >= 20 {
        "severe".to_string()
    } else if score >= 8 {
        "moderate".to_string()
    } else {
        "mild".to_string()
    }
}

/// Human-readable label for a severity band.
pub fn severity_label(level: &str) -> String {
    match level {
        "severe" => "Severe \u{2014} significant impact".to_string(),
        "moderate" => "Moderate \u{2014} targeted management advised".to_string(),
        "mild" => "Mild \u{2014} minimal impact".to_string(),
        _ => String::new(),
    }
}

/// CSS class for a severity band.
pub fn severity_class(level: &str) -> String {
    match level {
        "severe" => "severity-severe".to_string(),
        "moderate" => "severity-moderate".to_string(),
        "mild" => "severity-mild".to_string(),
        _ => String::new(),
    }
}

/// Run every SNOT-22 rule against the data, returning fired rules + totals.
pub fn calculate_snot22(data: &AssessmentData) -> (i32, SeverityLevel, i32, Vec<FiredRule>) {
    let mut fired_rules: Vec<FiredRule> = Vec::new();
    let mut total_score: i32 = 0;
    let mut answered_count: i32 = 0;

    for rule in all_rules() {
        let (score, answered) = (rule.evaluate)(data);
        if answered {
            answered_count += 1;
            total_score += score;
            fired_rules.push(FiredRule {
                id: rule.id.to_string(),
                category: rule.category.to_string(),
                description: rule.description.to_string(),
                score,
            });
        }
    }

    let severity_level = classify_snot22_score(total_score);
    (total_score, severity_level, answered_count, fired_rules)
}

/// Pure function: grades an otolaryngology assessment.
///
/// Returns a [`GradingResult`] containing the SNOT-22 total, severity band,
/// answered count, the per-question audit trail, and the clinical flag list.
pub fn grade(data: &AssessmentData) -> GradingResult {
    let timestamp = chrono::Utc::now().to_rfc3339();
    let (total_score, severity_level, answered_count, fired_rules) = calculate_snot22(data);
    let additional_flags = detect_additional_flags(data);

    GradingResult {
        total_score,
        severity_level,
        answered_count,
        fired_rules,
        additional_flags,
        timestamp,
    }
}
